### PR TITLE
[db-exporter] move resource-exporter cmd to db_exporter as subcommand and use statedb.dump_iter to replace statedb.dump

### PR DESCRIPTION
**Pull request by jiangying000**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

resource-exporter can export resources to csv
```
target/release/resource-exporter -r '0x1::Account::Balance<0x1::STC::STC>' --output ./xxdb0 --db-path ~/.starcoin/main/starcoindb/db/starcoindb --block-id 0x5b6c54192d69b230b443b013923b54b767561459fd558d590094f1c3fb630237  /token/value
```
```
real    0m4.122s
user    0m2.980s
sys     0m1.121s
```
Issue Number: #3428 

## What is the new behavior?

add `export-resource` to `starcoin_db_exporter`, same functionality as `resource-exporter`
```
./target/release/starcoin_db_exporter export-resource -r '0x1::Account::Balance<0x1::STC::STC>' --output ./xxdb9 --db-path ~/.starcoin/main/starcoindb/db/starcoindb --block-id 0x5b6c54192d69b230b443b013923b54b767561459fd558d590094f1c3fb630237 /token/value
```
```
real    0m6.893s
user    0m5.481s
sys     0m1.393s
```
- slower then before
-

### CODE REVIEW

File: cmd/db-exporter/src/main.rs [hunk 1]
```diff
@@ -1483,3 +1536,125 @@ pub fn apply_snapshot(
...
+        }
+        csv_writer.write_record(None::<&[u8]>)?;
+    }
+
+    let global_states_iter = statedb.dump_iter()?;
+    for (account_address, account_state_set) in global_states_iter {
+        let resource_set = account_state_set.resource_set().unwrap();
+
+        for (k, v) in resource_set.iter() {
+            let struct_tag = StructTag::decode(k.as_slice())?;
```
Comment (jiangying000) [inline cmd/db-exporter/src/main.rs:1591]: maybe linear scan take much time

Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1591]: time inside iterator loop:
original code took: 1267ms
this version takes: 5525ms

File: cmd/db-exporter/src/main.rs [hunk 2]
```diff
@@ -1483,3 +1536,125 @@ pub fn apply_snapshot(
...
+    {
+        csv_writer.write_field("address")?;
+        for f in fields {
+            csv_writer.write_field(f)?;
+        }
+        csv_writer.write_record(None::<&[u8]>)?;
+    }
+
+    let global_states_iter = statedb.dump_iter()?;
+    for (account_address, account_state_set) in global_states_iter {
```
Comment (jiangying000) [inline cmd/db-exporter/src/main.rs:1587]: time inside iterator loop:
original code took: 1267ms
this version takes: 5525ms

File: cmd/db-exporter/src/main.rs [hunk 3]
```diff
@@ -366,6 +378,34 @@ pub struct ApplySnapshotOptions {
...
+    #[clap(long, short = 'o', parse(from_os_str))]
+    /// output file, like accounts.csv
+    pub output: PathBuf,
+    #[clap(long, short = 'i', parse(from_os_str))]
+    /// starcoin node db path. like ~/.starcoin/barnard/starcoindb/db/starcoindb
+    pub db_path: PathBuf,
+
+    #[clap(long)]
+    /// block id (block hash) which snapshot at.
+    pub block_id: HashValue,
```
Comment (jiangying000) [inline cmd/db-exporter/src/main.rs:393]: could we use `block_hash` here ?

Comment (nkysg) [reply cmd/db-exporter/src/main.rs:393]: OK

File: cmd/db-exporter/src/main.rs (hunk 1 above)
Comment (nkysg) [reply cmd/db-exporter/src/main.rs:1591]: 测试的区块高度是多少？慢了几倍可以帮忙分析下原因吗

File: cmd/db-exporter/src/main.rs (hunk 2 above)
Comment (nkysg) [reply cmd/db-exporter/src/main.rs:1587]: what's the block height? could you help me to analyze the reason? If you are using a linux server, you can use the flamegraph see the difference

Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1587]: height is 6,881,689

Comment (nkysg) [reply cmd/db-exporter/src/main.rs:1587]: https://github.com/starcoinorg/starcoin/blob/6c0546e1326387d27b6c92cb3c449301dac229b9/cmd/db-exporter/src/main.rs#L458-L466

this is the related code

Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1587]: flamegraph

Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1587]: flamegraph of original `resource-exporter`
flamegraph-resource-exporter

Comment (nkysg) [reply cmd/db-exporter/src/main.rs:1601]: can you compare new commit and the original resource_exporter flamegraph

Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1601]: No clue yet, trying👀

Comment (nkysg) [reply cmd/db-exporter/src/main.rs:1601]: thx, I want to know reason. Let me think.

File: cmd/db-exporter/src/main.rs (hunk 1 above)
Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1620]: wrong guess, it took 198ms of 5300ms

File: cmd/db-exporter/src/main.rs (hunk 2 above)
Comment (jiangying000) [reply cmd/db-exporter/src/main.rs:1607]: just measured time as in 3808f98
turns out the iterator (`global_states_iter.next()`) took most of time, about 5000ms

The function corresponding to `global_states_iter.next` in flamegraph should be
`<starcoin_state_tree::state_tree::AccountStateSetIterator as core::iter::traits::iterator::Iterator>::next
`, inside which `rocksdb::DBImplReadOnly::Get(rocksdb::ReadOptions const&, rocksdb::ColumnFamilyHandle*, rocksdb::Slice const&, rocksdb::PinnableSlice*)` account for ~96% of time

File: cmd/db-exporter/src/main.rs [hunk 4]
```diff
@@ -1483,3 +1545,157 @@ pub fn apply_snapshot(
...
 }
+
+#[derive(Serialize, Debug)]
+pub struct AccountData<R: Serialize> {
+    address: AccountAddress,
+    #[serde(flatten)]
+    resource: Option<R>,
+}
+
+#[cfg(target_os = "linux")]
```
Comment (nkysg) [inline cmd/db-exporter/src/main.rs:1556]: 这行要去掉，要不只有linux平台定义这个函数了，其他平台无法编译

Comment (jiangying000) [reply cmd/db-exporter/src/main.rs]: ok 了

**Comment by nkysg** [comment]
要不先合并进去，性能那个问题几个issue再查下原因

**Review by nkysg — APPROVED** [review]

**Pull request merged**